### PR TITLE
Check addons versions before deployment

### PR DESCRIPTION
**Pull request by galal-hussein**

### CODE REVIEW

File: package/addons-update.sh [hunk 1]
```diff
@@ -65,7 +75,30 @@ for f in $(find /etc/kubernetes/addons -name '*.yaml'); do
...
+    "k8s-app=heapster,$HEAPSTER_IMAGE,/etc/kubernetes/addons/heapster/heapster"
+    "k8s-app=influxdb,$INFLUXDB_IMAGE,/etc/kubernetes/addons/heapster/influxdb"
+    "app=helm,$TILLER_IMAGE,/etc/kubernetes/addons/helm"
+   )
+
+# Check Addon version
+for i in "${addons_images[@]}"; do
+  current_version=$(kubectl get deployments -n kube-system -o=jsonpath="{..image}" -l "$(echo $i | cut -d"," -f1)" | cut -d" " -f1 | cut -d":" -f2)
+  desired_version=$(grep -r "$(echo $i | cut -d"," -f2)" /etc/kubernetes/addons | cut -d":" -f4)
+  if ! semver_gt ${current_version} ${desired_version}; then
```
Comment (alena1108) [inline package/addons-update.sh]: Nit: can we do the reverse comparison? usually looks better than !.

File: package/addons-update.sh [hunk 2]
```diff
@@ -65,7 +75,30 @@ for f in $(find /etc/kubernetes/addons -name '*.yaml'); do
...
+    "k8s-app=kube-dns,$KUBEDNS_IMAGE,/etc/kubernetes/addons/dns"
+    "k8s-app=grafana,$GRAFANA_IMAGE,/etc/kubernetes/addons/heapster/grafana"
+    "k8s-app=heapster,$HEAPSTER_IMAGE,/etc/kubernetes/addons/heapster/heapster"
+    "k8s-app=influxdb,$INFLUXDB_IMAGE,/etc/kubernetes/addons/heapster/influxdb"
+    "app=helm,$TILLER_IMAGE,/etc/kubernetes/addons/helm"
+   )
+
+# Check Addon version
+for i in "${addons_images[@]}"; do
+  current_version=$(kubectl get deployments -n kube-system -o=jsonpath="{..image}" -l "$(echo $i | cut -d"," -f1)" | cut -d" " -f1 | cut -d":" -f2)
```
Comment (alena1108) [inline package/addons-update.sh:108]: Do we take care of the case where we fail to fetch the image info by some reason.

Comment (galal-hussein) [reply package/addons-update.sh:108]: I added a check if the current version is empty to replace the addon anyway

**Comment by deniseschannon** [comment]
https://github.com/rancher/rancher/issues/10045

File: package/addons-update.sh [hunk 3]
```diff
@@ -65,7 +85,32 @@ for f in $(find /etc/kubernetes/addons -name '*.yaml'); do
...
+  sed -i "s|\$DASHBOARD_IMAGE|$DASHBOARD_IMAGE|g" ${f}
+  sed -i "s|\$KUBEDNS_IMAGE|$KUBEDNS_IMAGE|g" ${f}
+  sed -i "s|\$GRAFANA_IMAGE|$GRAFANA_IMAGE|g" ${f}
+  sed -i "s|\$HEAPSTER_IMAGE|$HEAPSTER_IMAGE|g" ${f}
+  sed -i "s|\$INFLUXDB_IMAGE|$INFLUXDB_IMAGE|g" ${f}
+  sed -i "s|\$TILLER_IMAGE|$TILLER_IMAGE|g" ${f}
+done
+
+addons_images=(
+    "k8s-app=kubernetes-dashboard,$DASHBOARD_IMAGE,/etc/kubernetes/addons/dashobard"
```
Comment (alena1108) [inline package/addons-update.sh]: typo "dashobard"

Comment (alena1108) [reply package/addons-update.sh]: see if you can parameterize /etc/kubernetes/addons to avoid repetition

**Pull request merged**